dotnet/src/InternalUtilities/connectors/AI/FunctionCalling/FunctionCallsProcessor.cs

```diff
@@ -190,62 +196,80 @@ public FunctionCallsProcessor(ILogger? logger = null)
+                var stringResult = ProcessFunctionResult(invocationContext.Result.GetValue<object>() ?? string.Empty);
+                return (stringResult, null, functionCall, invocationContext.Terminate);
+            }, cancellationToken);
+
+            // If concurrent invocation is enabled, add the task to the list for later waiting. Otherwise, join with it now.
+            if (functionTasks is not null)
             {
-                this.AddFunctionCallResultToChatHistory(chatHistory, functionCall, result: null, errorMessage: $"Error: Exception while invoking function. {e.Message}");
-                continue;
+                functionTasks.Add(functionTask);
             }
-            finally
+            else
             {
-                s_inflightAutoInvokes.Value--;
-            }
-
-            // Apply any changes from the auto function invocation filters context to final result.
-            functionResult = invocationContext.Result;
+                var functionResult = await functionTask.ConfigureAwait(false);
+                this.AddFunctionCallResultToChatHistory(chatHistory, functionResult.FunctionCall, functionResult.Result, functionResult.ErrorMessage);
 
-            object functionResultValue = functionResult.GetValue<object>() ?? string.Empty;
-
-            var result = ProcessFunctionResult(functionResultValue);
-
-            this.AddFunctionCallResultToChatHistory(chatHistory, functionCall, result);
+                // If filter requested termination, return last chat history message.
+                if (functionResult.Terminate)
+                {
+                    this._logger.LogDebug("Filter requested termination of automatic function invocation.");
+                    return chatHistory.Last();
+                }
+            }
+        }
 
-            // If filter requested termination, return last chat history message.
-            if (invocationContext.Terminate)
+        // If concurrent invocation is enabled, join with all the tasks now.
+        if (functionTasks is not null)
+        {
+            // Wait for all of the function invocations to complete, then add the results to the chat, but stop when we hit a
+            // function for which termination was requested.
+            await Task.WhenAll(functionTasks).ConfigureAwait(false);
+            foreach (var functionTask in functionTasks)
             {
-                if (this._logger.IsEnabled(LogLevel.Debug))
+                this.AddFunctionCallResultToChatHistory(chatHistory, functionTask.Result.FunctionCall, functionTask.Result.Result, functionTask.Result.ErrorMessage);
+
+                // If filter requested termination, return last chat history message.
+                if (functionTask.Result.Terminate)
                 {
                     this._logger.LogDebug("Filter requested termination of automatic function invocation.");
+                    return chatHistory.Last();
```

Comment:
@stephentoub, by the time the line 271 is executed, all functions have already been invoked. I think they should be added to the chat history unless there's a reason not to.
CC: @dmytrostruk

Reply:
I'm fine either way. I was trying to maintain behavior as close to the serial as possible, where once one iteration terminates nothing from subsequent calls are added. But there's also an argument to be made for adding already-computed results.